### PR TITLE
Spec03 $ref support

### PR DESCRIPTION
**Pull request by andreineculau**
Port of https://github.com/klarna/jesse/pull/34 by @jamesaimonetti

---

First pass at getting $ref support.

Punting on absolute URIs (or $ref value not starting with '#') by allowing the caller to pass a schema_loader_fun function that will return the schema JSON object, if found.

Appreciate any feedback!

### CODE REVIEW

File: src/jesse_state.erl [hunk 1]
```diff
@@ -35,11 +36,17 @@
...
+        , find_schema/2
         ]).
 
 -export_type([ state/0
              ]).
 
+%% Includes
+-include("jesse_schema_validator.hrl").
+
+-define(schema_loader_fun, fun jesse_database:read/1).
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Invalid macro name schema_loader_fun on line 48. Use UPPER_CASE.

File: src/jesse_validator_draft3.erl [hunk 2]
```diff
@@ -861,6 +864,35 @@ check_extends_array(Value, Extends, State) ->
              , Extends
              ).
 
+%% @private
+check_ref(Value, <<"#", LocalPath/binary>> = RefSchemaURI, State) ->
+  Keys = binary:split(LocalPath, <<"/">>, ['global']),
+  OriginalSchema = jesse_state:get_original_schema(State),
+
+  case local_schema(OriginalSchema, Keys) of
+    ?not_found -> handle_schema_invalid({'schema_unsupported', RefSchemaURI}, State);
```
Comment (gadgetci) [inline src/jesse_validator_draft3.erl]: According to **Elvis**:
> Line 873 is too long:     ?not_found -> handle_schema_invalid({'schema_unsupported', RefSchemaURI}, State);.

File: src/jesse_state.erl [hunk 3]
```diff
@@ -146,6 +161,20 @@ set_current_schema(State, NewSchema) ->
 set_error_list(State, ErrorList) ->
   State#state{error_list = ErrorList}.
 
+%% @doc Find a schema based on URI
+-spec find_schema(State :: state(), SchemaURI :: binary()) -> jesse:json_term() | ?not_found.
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 165 is too long: -spec find_schema(State :: state(), SchemaURI :: binary()) -> jesse:json_term() | ?not_found..

File: src/jesse_validator_draft3.erl [hunk 4]
```diff
@@ -861,6 +864,35 @@ check_extends_array(Value, Extends, State) ->
...
+  end;
+check_ref(Value, RefSchemaURI, State) ->
+  case jesse_state:find_schema(State, RefSchemaURI) of
+    ?not_found -> handle_schema_invalid({'schema_unsupported', RefSchemaURI}, State);
+    RefSchema -> check_ref_schema(Value, RefSchema, State)
+  end.
+
+%% @private
+check_ref_schema(Value, RefSchema, State) ->
+  TmpState = check_value(Value, unwrap(RefSchema), set_current_schema(State, RefSchema)),
```
Comment (gadgetci) [inline src/jesse_validator_draft3.erl]: According to **Elvis**:
> Line 884 is too long:   TmpState = check_value(Value, unwrap(RefSchema), set_current_schema(State, RefSchema)),.

File: src/jesse_validator_draft3.erl [hunk 5]
```diff
@@ -861,6 +864,35 @@ check_extends_array(Value, Extends, State) ->
...
+  Keys = binary:split(LocalPath, <<"/">>, ['global']),
+  OriginalSchema = jesse_state:get_original_schema(State),
+
+  case local_schema(OriginalSchema, Keys) of
+    ?not_found -> handle_schema_invalid({'schema_unsupported', RefSchemaURI}, State);
+    LocalSchema -> check_ref_schema(Value, LocalSchema, State)
+  end;
+check_ref(Value, RefSchemaURI, State) ->
+  case jesse_state:find_schema(State, RefSchemaURI) of
+    ?not_found -> handle_schema_invalid({'schema_unsupported', RefSchemaURI}, State);
```
Comment (gadgetci) [inline src/jesse_validator_draft3.erl]: According to **Elvis**:
> Line 878 is too long:     ?not_found -> handle_schema_invalid({'schema_unsupported', RefSchemaURI}, State);.

File: src/jesse_state.erl [hunk 6]
```diff
@@ -49,14 +56,12 @@
          , error_list         :: list()
          , error_handler      :: fun((#state{}) -> list() | no_return())
          , default_schema_ver :: atom()
+         , schema_loader_fun  :: fun((binary()) -> {ok, jesse:json_term()} | jesse:json_term() | ?not_found)
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 59 is too long:          , schema_loader_fun  :: fun((binary()) -> {ok, jesse:json_term()} | jesse:json_term() | ?not_found).

**Comment by andreineculau** [comment]
closing in favour of #5

**Pull request closed without merging**